### DISCUSSION TITLE
HxGrid is not updating rows after refreshing data from data provider

### DISCUSSION THREAD

**Discussion by prodev443** [havit/Havit.Blazor#940] [Q&A]
Hi

I have the following problem.

I need to update an HxGrid data with a button click, so I'm using the following code:

```
<HxButton Text="Actualizar grid" Color="ThemeColor.Primary" OnClick="RefreshGrid"/>

/// UI code and HxGridCode
<HxGrid @ref="Grid" TItem="Machine" DataProvider="GetGridData" PageSize="10" Responsive="true" InProgress="ProgressGridState">
/// Columns ...
</HxGrid>


@code
{
    private async Task<GridDataProviderResult<Machine>> GetGridData(GridDataProviderRequest<Machine> request)
    {
        /// I've verified that this method is executed on button click
        
        return new GridDataProviderResult<Machine>()
        {
            Data = await DataService.ListAsync(Filter, request.StartIndex, request.CancellationToken, request.Count ?? 10),
            TotalCount = await DataService.CountAsync(Filter, request.CancellationToken)
        };
    }

    private HxGrid<Machine> Grid;

    private async Task RefreshGrid()
    {
        ProgressGridState = true;

        await debouncer.DebounceAsync(1000, async (CancellationToken) =>
        {
            await Grid.RefreshDataAsync();
        });
        
        ProgressGridState = false;
        this.StateHasChanged();
    }
}
```

Can you help me please?

[Datagrid.zip](https://github.com/user-attachments/files/17769497/Datagrid.zip)

**Comment by hakenr** [comment]
Hi @prodev443,  

I can't reproduce the issue you're experiencing. For further analysis, please provide a runnable repro.  

BTW, You don't need to use `Debouncer` here:  
1. `HxGrid` includes built-in cancellation logic that cancels any ongoing `DataProvider` request when a new one is initiated.  
2. `HxButton` has built-in single-click protection, preventing a new click from being submitted until the previous `OnClick` handler finishes (this can be disabled using the `SingleSubmitProtection` parameter).

**Comment by prodev443** [comment]
Hi @hakenr ,

Thanks for your attendance and support.
I've created a [repository](https://github.com/prodev443/HavitGridBug/tree/master) with an  HxGrid in it. In the app you'll see the same behavior I've mentioned.
Also you can check this [YouTube video](https://youtu.be/kvmZ_fuLFWQ), it's a demonstration of the behavior.

Thanks in advance.

**Comment by hakenr** [comment]
Hi @prodev443,  
I finally found some time to look into your issue. Thanks for providing the repro code and recording the video.  

The issue isn't actually with `HxGrid` but rather with the repeated use of `DbContext` (in your case, `DataContext`). EF Core caches entities, so it's necessary to adjust `CustomerService` to avoid using the same `DataContext` instance for repeated queries.  

You can solve this by using `IDbContextFactory<>` like this:  

```csharp
public class CustomerService(IDbContextFactory<DataContext> contextFactory)
{
    private readonly IDbContextFactory<DataContext> _contextFactory = contextFactory;

    ...

    public async Task<IEnumerable<Customer>> ListAsync(CustomerFilter filter, int startIndex, int size = 10, CancellationToken cancellationToken = default)
    {
        using var context = await _contextFactory.CreateDbContextAsync(cancellationToken);
        var query = Filter(context.Customers.AsQueryable(), filter);

        return await query
            .Skip(startIndex)
            .Take(size)
            .ToListAsync(cancellationToken);
    }

    ...
}
```

In `Program.cs`, you'll need to update the registration like this:  

```csharp
builder.Services.AddDbContextFactory<DataContext>(
    options => options.UseSqlServer(connection)
);
```
